### ISSUE TITLE
Issues with -fPIC in upstream SDK

### ISSUE THREAD

**Issue by TomiBelan** [emscripten-core/emscripten#9690]
This C++ test program throws an exception and tries to catch it:
```cpp
#include <iostream>
int main() {
    std::cout << "hello\n";
    try {
        throw std::invalid_argument("foo");
    } catch (std::invalid_argument& e) {
        std::cout << "caught\n";
    }
    std::cout << "end\n";
}
```
Expected output: hello, caught, end.

After `./emsdk install latest`, `./emsdk activate latest`, `source ./emsdk_env.sh`:
```
$ em++ -c test.cc -o test.o && emcc -s DISABLE_EXCEPTION_CATCHING=0 -s EXCEPTION_DEBUG=1 -o test.js test.o && node test.js
hello
Compiled code throwing an exception, 5276944,26700,1
exception thrown: 5276944
```

The exception wasn't caught.

After `./emsdk install latest-fastcomp`, `./emsdk activate latest-fastcomp`, `source ./emsdk_env.sh`:
```
$ em++ -c test.cc -o test.o && emcc -s DISABLE_EXCEPTION_CATCHING=0 -s EXCEPTION_DEBUG=1 -o test.js test.o && node test.js
hello
Compiled code throwing an exception, 5277864,11312,522
can_catch on 5277864
  can_catch found 5277864,11312
cxa_begin_catch 5277864,stack,5277864
addref 5277864
caught
cxa_end_catch popped 5277864,5277864,stack,
decref 5277864
decref freeing exception 5277864,5277864,stack,
end
```
In fastcomp, the exception was properly caught.

According to `./emsdk list`, `latest` is currently equal to `1.39.0`.

**Comment by kripken** [comment]
I think this may be due to using wasm object files in the wasm backend, see the bullet point before last here: https://emscripten.org/docs/compiling/WebAssembly.html#backends

Basically, if you build `test.cc` to `test.o` then clang compiles the source into the final wasm right there. So you must say `-s DISABLE_EXCEPTION_CATCHING=0` at that stage as well, and not just during link. (In fastcomp, all codegen was done during link as object files contain LLVM IR.)

**Comment by TomiBelan** [comment]
Thank you for the ultra quick answer. You are correct! With `-s DISABLE_EXCEPTION_CATCHING=0` added to the `emcc` command line, it works.

Until now, I wasn't giving any special emscripten flags (`-s`, `--bind` etc) to the .c->.o compiler, only to the linker that creates the final .js+.wasm. Which flags should I give to the compiler (add to CXXFLAGS)? All `-s` flags?

Give me a moment - I'll try to verify if adding `-s DISABLE_EXCEPTION_CATCHING=0` to CFLAGS also helps with my original project, which had different broken symptoms, and not just the minimal test case above.

**Comment by TomiBelan** [comment]
Adding `-s DISABLE_EXCEPTION_CATCHING=0` was necessary but not sufficient to solve my original issue. There is something else wrong with "upstream". But I haven't had any luck yet with producing a minimal test case. It's not a very good bug report but I'll do my best to describe it.

I have a project that consists of multiple C++ files, and statically links a third party library which is built with slightly different flags (e.g. `-I`, `-std=`). One file from that library contains a global constructor which is throwing bizarre exceptions (hence why I thought exceptions are broken). When I tried adding `std::cerr << "some debug string";` to that constructor, it actually printed a fragment of another string constant from elsewhere in that file. So I'm guessing there is some weird linker issue where the read-only-data section of this .o file is placed at some incorrect offset, and this is causing all sorts of nonsensical errors.

With `-s WASM_OBJECT_FILES=0`, the problem disappears. Do you have any ideas what could be causing this bug or how can I troubleshoot it?

**Comment by kripken** [comment]
@TomiBelan That does sound like a weird bug, maybe in the wasm object file support, and no known issues there, so this is worrying... would be good to reduce this as much as you can and provide a testcase, if that's possible. As the weirdness happens in a global constructor, maybe it's enough to focus on just that one library?

One thing that may help is to use ASAN or the other sanitizers, which [do work in upstream](https://emscripten.org/docs/debugging/Sanitizers.html). Running them both with and without object files may be helpful as maybe there some undefined behavior.

@sbc100 any ideas on debugging an issue like this that only happens with wasm object files?

**Comment by TomiBelan** [comment]
I'll do my best to make a test case, but it might take a while. Meanwhile: can you tell me a bit about what is a "wasm object file" and how can I inspect or disassemble its contents? I noticed I can use emsdk/upstream/bin/wasm2wat to read a .wasm file, but what tools work on .o? I'm not very familiar with this area, but I'd like to figure out if the .o file is already wrong or if it's the linker's fault.

**Comment by TomiBelan** [comment]
I've got it! It is because of `-fPIC`.

I guess fastcomp ignored `-fPIC` and upstream doesn't really support it?

The minimal test case is:
```cpp
#include <iostream>
void printb();
int main() {
    std::cerr << "main\n";
    printb();
    std::cerr << "end\n";
}
```
```cpp
#include <iostream>
void printb() {
    std::cerr << "here in b\n";
}
```
```shell
$ em++ -c a.cc && em++ -c b.cc && emcc -o ab.js a.o b.o && node ab.js 
main
here in b
end
$ em++ -c a.cc && em++ -c b.cc -fPIC && emcc -o ab.js a.o b.o && node ab.js 
main
IcEEend
```

To answer my own question, `wasm-objdump -d` and `wasm-objdump -x` work pretty well, and support both .wasm and .o files.

When -fPIC is enabled, printb adds `global.get 0 <env.__memory_base>` (which equals 1024) to a i32.const string offset. But the string offset is already counted from 1024. (In my particular build of ab.wasm, "main\n" is at 1024, "end\n" at 1030 and "here in b\n" at 1035.) Adding 1024 a second time produces the wrong address.

The wasm file has three data segments, and the string constants are at the beginning of the first data segment.
```
Data[3]:
 - segment[0] size=13436 - init i32=1024
 - segment[1] size=684 - init i32=14464
 - segment[2] size=5800 - init i32=15152
```
Since data starts at 1024, setting __memory_base to 1024 doesn't seem so unreasonable. Most likely, a "real" fix would be to either set __memory_base to 0, or to make the linker stop adding 1024 to the i32.const offset, but I don't know which is better.

-----

In conclusion, after I added `-s DISABLE_EXCEPTION_CATCHING=0` and removed `-fPIC`, my project works with both fastcomp and upstream. (Now I just need to figure out how to do it without hackishly patching that third party library's Makefile.)

My problem is solved, but maybe there are things that could be done to prevent it, such as warning if people use different `-s` settings, or warning on use of `-fPIC`, or fixing `-fPIC`. Feel free to close this issue, or feel free to keep it open to track those ideas.

**Comment by kripken** [comment]
Thanks @TomiBelan , excellent debugging and great testcase!

Yes, this looks like a bug in upstream's `-fPIC` support, that definitely should not happen. Adding to blocker list.

**Comment by kripken** [comment]
(Btw, out of curiosity I verified `-fsanitize=undefined` does show a bad memory usage in `__wasm_call_ctors` on the testcase, but it does not give any indication as to `-fPIC` being involved, so the help it can give for stuff like this is limited.)

**Comment by manumartin** [comment]
Hi,

I'm also seeing a weird behaviour with -fPIC, a simple cout << "hello world" program works without -fPIC but not with it. With -fPIC it builds and executes but prints weird characters.

It happens with emscripten 1.38.47 + LLVM master:

#include <iostream>

int main() {
        std::cout << "Hello worlds dfa sdf asd fasfd sadf " << std::endl;
}

em++ -o index.html -fPIC main.cpp

**Comment by pmp-p** [comment]
same there on micropython/cpython with C, `-fPIC` will corrupt any build and with `-fPIC -s MAIN_MODULE=1` the `Fatal: GOT.func entry with no import/export: $emscripten_longjmp_jmpbuf` is back anyway.

**Comment by sbc100** [comment]
The longjmp issues is #2180 and the fix is in https://github.com/WebAssembly/binaryen/pull/2382

**Comment by sbc100** [comment]
I re-opened the original issue: #9013.   Closing this as a dupe of that.
